util/gpgmeutils.c

```diff
@@ -358,6 +358,127 @@ find_email_encryption_key (gpgme_ctx_t ctx, const char *uid_email)
     }
 }
 
+/**
+ * @brief Wrapper for fread for use as a GPGME callback.
+ *
+ * @param[in]  handle   The file handle.
+ * @param[out] buffer   The data buffer to read data into.
+ * @param[in]  size     The size of the buffer.
+ *
+ * @return The number of bytes read or -1 on error.
+ */
+static ssize_t
+gvm_gpgme_fread (void *handle, void *buffer, size_t size)
+{
+  int ret;
+  FILE *file = (FILE *)handle;
+
+  ret = fread (buffer, 1, size, file);
+  if (ferror (file))
+    return -1;
+  return ret;
+}
+
+/**
+ * @brief Wrapper for fread for use as a GPGME callback.
+ *
+ * @param[in]  handle   The file handle.
+ * @param[in]  buffer   The data buffer to read data into.
+ * @param[in]  size     The amount of buffered data.
+ *
+ * @return The number of bytes written or -1 on error.
+ */
+static ssize_t
+gvm_gpgme_fwrite (void *handle, const void *buffer, size_t size)
+{
+  int ret;
+  FILE *file = (FILE *)handle;
+
+  ret = fwrite (buffer, 1, size, file);
+  if (ferror (file))
+    return -1;
+  return ret;
+}
+
+/**
+ * @brief Create a GPGME data buffer with custom read and write functions.
+ *
+ * This is neccessary as gpgme_data_new_from_stream may cause problems
+ *  when trying to write to the stream after some operations.
+ *
+ * @param[out]  new_data  The new GPGME data buffer.
+ * @param[in]   file      The stream to read from and write to.
+ *
+ * @return The return value from gpgme_data_new_from_cbs.
+ */
+static gpgme_error_t
+gvm_gpgme_data_new_from_stream (gpgme_data_t *new_data, FILE *file)
+{
+  struct gpgme_data_cbs *callbacks;
+
+  callbacks = malloc (sizeof (struct gpgme_data_cbs));
+  memset (callbacks, 0, sizeof (struct gpgme_data_cbs));
+  callbacks->read = gvm_gpgme_fread;
+  callbacks->write = gvm_gpgme_fwrite;
+
+  return gpgme_data_new_from_cbs (new_data, callbacks, file);
```

Comment:
Where is callbacks freed?  Should callbacks->release be set to do this?

Reply:
Alternatively callbacks could be passed in from the caller, so it doesn't need to be freed.

Reply:
Done in 8428f83.